Fixes: #10404 Adapting preprocessing to be able to process inputs with NaN

pinakinathc opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes: #10404 
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.

Changed utils : **validation.py** : check_array : to raise an Error only when there is numpy.inf and NOT numpy.nan
utils : **extmath.py** : _incremental_mean_var : to calculate mean using numpy.nanmean() when there is a numpy.nan present, for variance used numpy.nanvar()

#### Any other comments?
@jnothman and others please check whether this is okay or not. I might have understood it wrong (once again)

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

pinakinathc committed:
[WIP] Fixes: 10404 Adapting preprocessing to be able to process inputs with NaN

Changed utils : **validation.py** : check_array : to raise an Error only when there is numpy.inf and NOT numpy.nan
utils : **extmath.py** : _incremental_mean_var : to calculate mean using numpy.nanmean() when there is a numpy.nan present, for variance used numpy.nanvar()

Signed-off-by: Pinaki Nath Chowdhury <pinakinathc@gmail.com>

jnothman commented:
Please follow the contributor guidelines. Your PR description should mention which issue it fixes, with "Fixes" before it if appropriate. You should ideally have a task list in your WIP PR description so we know what you think you've done and what you think remains. And you should have consulted with @ericchang00 on #10404 to divide up the work.

ericchang00 commented:
@pinakinathc Just a note, you'll want to add unit tests for the changes you're making. Also, it looks like these changes broke quite a few tests in unrelated libraries. I'd suggest adding NaN handling in `sklearn/preprocessing/data.py`.

jnothman commented:
I don't mind handling it in extmath, but it'll have to be done carefiully

On 9 January 2018 at 12:39, Eric Chang <notifications@github.com> wrote:

> @pinakinathc <https://github.com/pinakinathc> Just a note, you'll want to
> add unit tests for the changes you're making. Also, it looks like these
> changes broke quite a few tests in unrelated libraries. I'd suggest adding
> NaN handling in sklearn/preprocessing/data.py.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10432#issuecomment-356152751>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz68nbzdCJSEa_lfYMCwg82YREDO0Cks5tIsNWgaJpZM4RXLkZ>
> .
>

pinakinathc committed:
Revert "Fixes: #10404 Adapting preprocessing to be able to process inputs with NaN"

This reverts commit 751514210974ecf35de93d550166e12465f13b32.

pinakinathc committed:
Fixes: #10404 modified utils: sparsefunctions_fast.pyx:
Functions like _csr_mean_variance_axis0 and _csc_mean_variance_axis0 can process NaN

- [x] StandardScalar can process NaN in cases of sparse matrix without increment
- [ ] StandardScalar can process NaN in cases of sparse matrix with increments
- [ ] StandardScalar can process NaN in cases of non-sparse matix without increments
- [ ] StandardScalar can process NaN in cases of non-sparse matrix with increments

Signed-off-by Pinaki Nath Chowdhury <pinakinathc@gmail.com>

jnothman commented on sklearn/utils/sparsefuncs_fast.pyx:
```diff
@@ -184,7 +193,10 @@ def _csc_mean_variance_axis0(np.ndarray[floating, ndim=1] X_data,
         counts = endptr - startptr
 
         for j in xrange(startptr, endptr):
+            if X_data[j] == np.nan:
```
Comparison to nan doesn't work. You need to use isnan, or look for a C implementation to use in Cython

jnothman reviewed (commented):
Please start by adding test cases

pinakinathc commented:
@jnothman utils: sparsefunctions.py: line [143](https://github.com/scikit-learn/scikit-learn/blob/b94876a3a20e1d629698ea1c7b358169b97bfcce/sklearn/utils/sparsefuncs.py#L143) and in line [150](https://github.com/scikit-learn/scikit-learn/blob/b94876a3a20e1d629698ea1c7b358169b97bfcce/sklearn/utils/sparsefuncs.py#L150)
if **X** is a CSR or CSC sparse matrix
the either way there is not much difference, i mean the rest of the **if** and **elif** is same.

Just confirming. we can add them in 1 **if** statement with an **or** clause
